### DISCUSSION TITLE
RP2040 in Amiga 3000 - Boot delay

### DISCUSSION THREAD

**Discussion by SezaroSteve** [ZuluSCSI/ZuluSCSI-firmware#138] [General]
Hello,

I have installed a RP2040 with Firmware 1.2.0 into my Amiga 3000. When the "SCSI Timeout" on the onboard SCSI-Controller is set to the default of "Short", the system does not detect the ZuluSCSI. In other words, the "insert floppy animation" is shown instead. When I then reboot (not power-cycle) with Ctrl-Amiga-Amiga, the system detects and boots from the ZuluSCSI just fine.

When I change the SCSI Timeout to "Long", Then power-on the Amiga, there is a delay and then the system boots up normally.
This delay is quite long. But it's better than the "short" setting which prevents boot-from-scsi after power-on.

The thing is that his "long" or "short" setting (stored in the SCSI-controllers NVRAM) was used with spinning harddrives. Drives that where slow to report themselves on the SCSI-Bus where given more time to spin up and make their presence known.
Ofcourse, with modern flash, this is moot. But the "Short" setting is a tad too aggressive. It is so fast, that the ZuluSCSI is not seen on the bus yet.

Now to my question: in the  zuluscsi.ini  there is a parameter called "SelectionDelay = 255" (this is the default). What does this parameter do exactly?
Does it control the delay when the Zulu reports it's presence on the SCSI-bus? Or is it used to delay virtualized SCSI-Devices to appear on the bus? Or something else entirely?

**Comment by PetteriAimonen** [comment]
Selection delay is delay to responding to pulse on `SEL` signal. The default setting `255` means automatic and selects between 0 and 1 ms depending on quirks setting. It has no effect on boot delay.

Currently the boot delay on RP2040 is normally 1 - 2 seconds depending on number of images. There is a file `zululog.txt` written on the SD card, it will show how long it takes for you.

Do you know how short the "short" delay is, i.e. how much faster the boot would need to be?

**Reply by SezaroSteve** [reply]
No, but I will ask on EAB.

**Comment by SezaroSteve** [comment]
Ok I got it. A member on EAB found this in the documentation from a tool that is used to configure the A3000's onboard controller and store the values in its NVRAM.

> SCSI_TIMEOUT
> Adjusts the timeout value for SCSI device selection.
> A value of 0 will produce short timeouts (128 ms).
> A value of 1 produces long timeouts (2 sec). This is used
> for Seagate drives (and some Maxtors apparently) that
> don`t respond to selection until they are fully spun up
> and intialised.

So "short" is 128 ms.

Idea: make it configurable in the .ini
Ranging from "0 = no timeout" (as fast as possible) to 20000 ms (2s seconds).
The default minimum could be 1 second like it is now.

When the A3000 is powered on, the Zulu is booting it's firmware. At the same time, the onboard SCSI controller initializes itself and when finished, waits 128 ms.
If the Zulu is ready to go fast enough. Faster than the end of the "scsi controller initialize + 128 ms" cycle, it might just work with the A3000 default timeout setting "short". If all 7 scsi-ID's are used (in my A1200 I use 6 virtual HDD's and 1 virtual CD-ROM drive) and having so many devices causes the Zulu to be "late" (longer than "scsi controller initialize + 128 ms") than yeah, there is nothing that can be done. Then the user must configure "long" and so be it.
But for simple setup's, with one or two virtual SCSI devices, maybe the Zulu is finished quick enough to stay under "scsi controller initialize + 128 ms" and then it could work with "short".

**Reply by PetteriAimonen** [reply]
Yeah, it is a question on how fast the firmware can initialize the SD card and gather data about the images. Could probably be optimized a bit. And because the SCSI IDs are not known before processing the images, it cannot respond to selection pulses early.

**Reply by MeTalOneOEight** [reply]
Can't you use the ROM drive in the flash for that purpose?
Or add a cached SD card setup for the people who use the same SD Content all the time?

**Reply by PetteriAimonen** [reply]
Many things are possible, it's just a question of what makes the most sense & is worth the development time.
Optimizing the SD card initialization is the most straightforward and has benefit for all users, without impacting ability to change SD cards etc.

**Reply by SezaroSteve** [reply]
> Can't you use the ROM drive in the flash for that purpose?
> Or add a cached SD card setup for the people who use the same SD Content all the time?

It will **always** be necessary to validate the "SD configuration cache" contents with the SD-Card. If a user swaps SD-Cards, or puts different images on it, add stuff, funky, bad things could happen.
Can't blindly rely on a configuration-cache like that. How can the RP know if the cache-contents are correct. And when not, it needs to rebuild the cache.
Thus what is the point of having such a cache when the verification (and possible cache rebuild) takes more time than just:
- "read the card and process the images" (2 actions)
versus;
- "read the card and process the images, validate the cache" (3 actions)
or even worse:
- "read the card and process the images, validate the cache, discovering it's not valid and rebuild it" (4 actions).

Blindly assuming that the SD-Card's contents are exactly as what is written in the configuration-cache, without any checks (to save time), is not a good idea.

The only thing that might work. is to add a setting in the .ini file saying "do not verify the cache, just assume it's correct = 1" (the default would be "0", at which point the user has full responsibility and if he/she/it changes the content of the card and leave that setting to "1" thus causing problems.

Having said that, I agree with PetterAimonen that optimizing the SD card initialization is the most logical approach. If the RP2040 does this as fast as it possibly can, and the Amiga 3000's onboard SCSI is still too fast with its "short" setting, then we A3000 owners will simply have to live with the "long" setting. So be it.

**Reply by PetteriAimonen** [reply]
I now realized two quite simple to fix time wasters:

1. The LED blink takes 500 ms, just sleeping.
2. Every .ini setting read re-opens the ini file. The file contents could be cached in RAM on boot / SD card hotplug.

**Reply by SezaroSteve** [reply]
Ok then i'll gladly wait for Firmware v1.2.1 then ;-)
I'll beta-test for you.

(I'll need to wait for my Amiga 3000 to come from the repair-shop though so can't test this week).

**Reply by PetteriAimonen** [reply]
There is a test build here now (download zip at bottom of page):
https://github.com/ZuluSCSI/ZuluSCSI-firmware/actions/runs/4133419324

For me it finishes the initialization in 240 ms.
Hopefully that is fast enough, all the easy optimizations have been done now.

**Reply by SezaroSteve** [reply]
Tnx Petter, will try after my A3000 comes back from repairs. I expect her to be in the hospital for another week.
Is that timing the only change?

**Comment by SezaroSteve** [comment]
Hello Petteri,

My A3000 is back from repairs. The ZuluSCSI in there has firmware 1.2.1. In order to try your modification/optimization, can I simply overwrite 1.2.1 with the firmware in that link?
I assume that this modification is not in v2023.03.14 ?

**Reply by PetteriAimonen** [reply]
Yeah, it hasn't been merged yet. I would like to know whether it actually boots fast enough with those modifications, as I don't have an Amiga for testing.

**Comment by SezaroSteve** [comment]
It works perfectly :-)
I tested 10 times, letting the A3000 boot cold. I made sure the capacitors where drained from any leftover electron lurking inside them to ensure a cold boot is really as cold as it can be (without sending it to the northpole).
10 tests -> 10 perfect cold starts.

You did it !  :-)

Idea: make this optimisation optional. I cannot foresee the impact of this on other systems. What it might break so to speak. So maybe "Default is disabled" and setting a parameter to "1" in the zuluscsi.ini  activates it. This is what A3000 users should do. It could even be mentioned in the .ini file's comments like "Amiga 3000 owners, or other systems that need the initialisation-speedup, should set this to 1". Something like that.
If this optimisation is completely harmless, then all of this is not needed and it can just be active as standard.


There are 6 Images on the card which take longer to scan than just 1 image and it still works flawlessly.
This is my A3000D's zuluscsi.ini  in case it is of interest:
_________________________________________
```
[SCSI]
Debug = 0   # Same effect as DIPSW2, enables verbose log messages
SelectionDelay = 255   # Millisecond delay after selection, 255 = automatic, 0 = no delay
PhyMode = 0   # 0: Best available  1: PIO  2: DMA_TIMER  3: GREENPAK_PIO   4: GREENPAK_DMA
Quirks = 0   # 0: Standard, 1: Apple, 2: OMTI, 4: Xebec, 8: VMS
EnableUnitAttention = 0 # Post UNIT_ATTENTION status on power-on or SD card hotplug
EnableSCSI2 = 1 # Enable faster speeds of SCSI2
EnableSelLatch = 0 # For Philips P2000C and other devices that release SEL signal before BSY
MapLunsToIDs = 0 # For Philips P2000C simulate multiple LUNs
MaxSyncSpeed = 10 # Set to 5 or 10 to enable synchronous SCSI mode, 0 to disable

[SCSI0]
Vendor = "QUANTUM"
Product = "LPS270S"
Version = 0E4

[SCSI1]
Vendor = "QUANTUM"
Product = "ST21S012"
Version = 02-C

[SCSI2]
Vendor = "QUANTUM"
Product = "ST21S012"
Version = 02-C

[SCSI3]
Vendor = "QUANTUM"
Product = "ST21S012"
Version = 02-C

[SCSI4]
Vendor = "Seagate"
Product = "ST3600N"
Version = A06

[SCSI6]
Vendor = "Pioneer"
Product = "DR-506S"
Type = 2
IMG0 = MacOS8_1.iso
```

ID0 = HD0 = 200MB
ID1 = HD1 = 2 GB
ID2 = HD2 = 2 GB
ID3 = HD3 = 2 GB
ID4 = HD4 = 500 MB

The virtual CD-ROM drive is ID6.
I skip ID5 in the Zulu as that is used by the external Iomega ZIP drive which is permanently attached to the external SCSI connector of this A3000 Desktop.

Some background info:
The bus is terminated externally with an active terminator on the ZIP drive. The ZIP drive has it's own internal termination turned off. This allows correct operation of the bus when the ZIP drive is not powered on (I don't use it very often).

That external active terminator is the "GTERM25 - DB25 Active SCSI Terminator with diagnostics LEDs" from GGlabs. I highly recommend this solution to anyone as one can immediately see if the entire SCSI-bus is properly terminated and if there are any problems just by looking at the four diagnostics LED's on the terminator.

**Reply by PetteriAimonen** [reply]
Thanks for testing, I've opened pull request #159 and this will be part of the next release.
